web/packages/teleterm/src/ui/services/connectionTracker/trackedConnectionOperationsFactory.ts

```diff
@@ -154,6 +158,55 @@ export class TrackedConnectionOperationsFactory {
     };
   }
 
+  private getConnectionGatewayAppOperations(
```

Comment:
We might be able to reuse `getConnectionGatewayOperations` here instead of adding a new function, no?

Reply:
By that I mean that in contexts where we can treat a gateway as a gateway with a generic target, we should do so. After applying the two suggestions mentioned above, `getConnectionGatewayAppOperations` becomes pretty much identical to `getConnectionGatewayOperations`. This suggests that we should be able to use `TrackedGatewayConnection` for both databases and apps. `targetUri` could be then used to differentiate between them where needed, for example in the UI when saying "DB" vs "APP".

Reply:
That's right, those gateway types are really similar. Probably we could go even further and use a single type for a document? 🤔
I'm only thinking about the backward compatibility, what will happen if an older version of Connect will try to open an app connection treating it like a db? (probably nothing good)
OTOH, now the app just crashes when it see an unsupported connection, I don't know what is worse :)

Reply:
> That's right, those gateway types are really similar. Probably we could go even further and use a single type for a document? 🤔

I wasn't thinking too much about documents because they have considerably different UI, especially Kube vs the rest. OTOH there's a lot of similar actions (like changing the port) which could be refactored to a hook or reused in some other way between db and app gateways (I haven't looked at the UI code for app gateways in this PR yet).

In case of connections, the only meaningful difference between is the label shown in the list, or at least that's what I thought until you mentioned backward compatibility.

> I'm only thinking about the backward compatibility, what will happen if an older version of Connect will try to open an app connection treating it like a db? (probably nothing good)

That's a good shout. It's more like forward compatibility I suppose. It's worth checking out, but I think besides UI inconsistencies nothing should make the app unusable? The Electron app will attempt to create a gateway for the given target URI, but tshd should fail at either parsing the target URI (if the given version of tshd supports kube gateways where we inspect if the target URI is of a db or a kube cluster) or finding the database.

Reply:
Right now if I go back to the previous version of the app which doesn't support app gateways, I get:

```
TypeError: Cannot destructure property 'rootClusterUri' of 'this._trackedConnectionOperationsFactory.create(...)' as it is undefined.
    at eval (webpack-internal:///./src/ui/services/connectionTracker/connectionTrackerService.ts:165:9)
    at Array.map (<anonymous>)
    at ConnectionTrackerService.getConnections (webpack-internal:///./src/ui/services/connectionTracker/connectionTrackerService.ts:162:35)
    at useConnections (webpack-internal:///./src/ui/TopBar/Connections/useConnections.ts:36:35)
    at Connections (webpack-internal:///./src/ui/TopBar/Connections/Connections.tsx:52:86)
    at renderWithHooks (webpack-internal:///../../../node_modules/react-dom/cjs/react-dom.development.js:16305:18)
    at mountIndeterminateComponent (webpack-internal:///../../../node_modules/react-dom/cjs/react-dom.development.js:20069:13)
    at beginWork (webpack-internal:///../../../node_modules/react-dom/cjs/react-dom.development.js:21582:16)
    at HTMLUnknownElement.callCallback (webpack-internal:///../../../node_modules/react-dom/cjs/react-dom.development.js:4164:14)
    at Object.invokeGuardedCallbackDev (webpack-internal:///../../../node_modules/react-dom/cjs/react-dom.development.js:4213:16)
```

Reply:
I decided to remove `doc.gateway_app` and  `connection.app` types. They were mostly copies of our regular document gateway and the gateway connection.  It is much better to simply differentiate them by their URI. 

If we ever need to add more properties to the app gateways, we can always create a new type.

Another advantage is that it fixes the errors that show up after downgrading the app. In case of connection, it will be shown as a db connection. If you try to connect you will see this nice error:
<img width="711" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/5ea1c7e0-5217-4677-bb3b-b355f29ebe9c">

The only thing that I have to fix, is to use `routing.parseClusterUri` instead of `routing.parseDbUri` [here](https://github.com/gravitational/teleport/blob/77d95253b260a0a8c3ff7807e5e23ee3c07c8b64/web/packages/teleterm/src/ui/services/connectionTracker/trackedConnectionOperationsFactory.ts#L106), since it can be also an app uri (we could use `routing.parseClusterUri` from the beginning, we don't even need anything from db at that point).
I will create a PR and backport it.